### ISSUE TITLE
Better control over style guide structure for patterns

### ISSUE THREAD

**Issue by tylersticka** [cloudfour/drizzle-builder#16]
Overall, I'm happy with our Fabricator-based pattern libraries. But possibly my biggest complaint is how _gargantuan_ the "Components" section gets. I look at stuff like [Solid](http://solid.buzzfeed.com/) and miss the days of being able to have more focused pages.

I know you can _theoretically_ do that with Fabricator, but it involves creating additional views for every single category, which just seems silly when they all need to do the exact same thing.

My dream structure for patterns/material might look like this (no idea how feasible it is):

```
├── typography/          → /typography.html
│   ├── headings.hbs     → /typography.html#headings
│   └── paragraphs.hbs   → /typography.html#paragraphs
└── components/
    ├── button/          → /button.html
    │   ├── default.hbs  → /button.html#default
    │   └── primary.hbs  → /button.html#primary
    └── field/           → /field.html
        ├── text.hbs     → /field.html#text
        └── search.hbs   → /field.html#search
```

In this example, the `components` directory is purely a means of denoting what non-linked heading a group of style guide patterns should be confined to.

As with all my feature requests so far, I'm open to this being managed less in filenames and more in Frontmatter. The end result is more important to me than how we get there.

(This may relate to another issue I created, #14. I can see a universe where we have a directory of pages to be directly converted as-is, and a directory for things we want available to views as an array or object.)

**Comment by erikjung** [comment]
I like this idea. @lyzadanger and I chatted earlier about ways to eliminate the need to manually create HTML files that correspond with each collection. How do we feel about having _no_ components overview page? Is this example implying that `components.html` would not exist?

**Comment by tylersticka** [comment]
> Is this example implying that components.html would not exist?

Exactly. I basically want to ~~steal~~ borrow what [Solid](http://solid.buzzfeed.com/) does:

![screen shot 2016-03-07 at 4 32 27 pm](https://cloud.githubusercontent.com/assets/69633/13588171/8df13d18-e482-11e5-9ee9-1d8247381f6d.png)

**Comment by lyzadanger** [comment]
@tylersticka I'll probably need some hand-holding through this as I recall that the question of "what levels should get a generated page?" was thorny in the past. But it should be doable in some manner!

**Comment by tylersticka** [comment]
> in some manner!

I know I said this in my original description, but I really am open to any way of doing this. I gave a very specific example, but if this doesn't happen on the file system... if this is something within frontmatter, or settings, or if you need to create some sort of `index` file in directories that you want to compile... I'm completely open.

**Comment by lyzadanger** [comment]
@tylersticka et al

One of the things that bugs me about the current implementation is that there is a hard limit on depth of "collections". There is really only one level of depth (there are "collections" and "sub-collections" and that is where the story ends). 

As part of extending this and making it more flexible, I'd propose that we refactor this part of the code to use a recursive approach and thus not limit us to depth. We can then consider what the best defaults are for levels and groupings and whatnot. But I think that having a non-limited data structure is a good first step.

I happen to be in this area of code right now, so quick feedback on this particular thing is much appreciated :).

**Comment by tylersticka** [comment]
In theory, I agree. In practice, if we can get the navigation structure I described in a sane, intuitive way, I'll be happy with however that happens.

**Comment by lyzadanger** [comment]
This might be easier to grok with an example. Riffing on @tylersticka's from above:

```
├── typography/          → /typography.html
│   ├── headings.hbs     → /typography.html#headings
│   └── paragraphs.hbs   → /typography.html#paragraphs
└── components/
    ├── button/          → /button.html
    │   ├── default.hbs  → /button.html#default
    │   └── primary.hbs  → /button.html#primary
    └── field/           → /field.html
        ├── text.hbs     → /field.html#text
        └── search.hbs   → /field.html#search
```

This structure happens to follow what is already the codified depth. You'd end up with a data structure that would be keyed something like:

```
- typography
- components
    - button
    - field
```

But let's say we have a dir structure that looks like (this is totally hypothetical):

```
├── typography/          → /typography.html
│   ├── headings.hbs     → /typography.html#headings
│   └── paragraphs.hbs   → /typography.html#paragraphs
└── components/
    ├── button/          → /button.html
    │   ├── default.hbs  → /button.html#default
    │   └── primary.hbs  → /button.html#primary
     ├──── components/
      |                  ├── default.hbs  → ???
     ├──── foobar/
      |                  ├── ding.hbs  → ???
    └── field/           → /field.html
        ├── text.hbs     → /field.html#text
        └── search.hbs   → /field.html#search
```

In the current architecture, I believe you'd end up with

1. A namespace collision on `components`
2. No collection defined for the `components` under `button`
3. I do not believe you'd get an entry for `foobar`, either

You'd still end up with a similar data object to above

```
- typography
- components
    - button
    - field
```

Though without writing some tests I'm not certain what the `components` collection would contain. 

In my proposed refactor, the structure would follow the dir structure more closely, e.g.:

```
- typography
- components
    - button
         - components
         - foobar
    - field
```

To summarize tl;dr, I'd like to make the underlying data structure represent the file structure more. _Then_ we can decide how to handle different levels of groupings. Note that what I'm talking about doesn't yet get to the `.html` output, but how to organize the _data_ to make generating those in the most flexible way possible.

**Comment by tylersticka** [comment]
Sure, that makes sense.

**Comment by lyzadanger** [comment]
> My dream structure for patterns/material might look like this (no idea how feasible it is):

Good news. I think we're damn close to exactly that. I forgot about the in-page anchors—I can totally add those.

**Comment by erikjung** [comment]
This is done-ed.
